ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/ParticipantTestContext.scala

```diff
@@ -642,6 +665,14 @@ private[testtool] final class ParticipantTestContext private[participant] (
       SetTimeModelRequest(nextSubmissionId(), Some(mrt.asProtobuf), generation, Some(newTimeModel)),
     )
 
+  def prune(pruneUpTo: Option[LedgerOffset], attempts: Int): Future[Empty] =
+    eventually(attempts = attempts, runAssertion = {
```

Comment:
I will add the following comment:

```
  def prune(pruneUpTo: Option[LedgerOffset], attempts: Int): Future[Empty] =
+    // Distributed ledger participants need to reach global consensus prior to pruning. Hence the "eventually" here:
    eventually(attempts = attempts, runAssertion = {
      services.participantPruning.prune(PruneRequest(pruneUpTo, nextSubmissionId()))
    })
```

Reply:
Discussed online: some ledgers need to perform ledger-wide synchronization before allowing participants to delete local data. I would suggest one of the following:

- Require `WriteParticipantPruningService.prune` implementations to wait a resonable time until they are ready to prune. This would remove the need for `eventually`. On the other hand, the ledger API call could take a long time and possibly trigger some network timeout.
- Clearly document in `participant_pruning_service.proto` and in `PruningResult.scala` which error code means that the participant is not ready yet and that the client application should retry later at the same offset. You could even consider adding a `PruningResult.WorkingOnItPleaseRetryLater` case for this.

Speaking of which, the protobuf file could use comments on what error messages the prune call can return, see the [command service example](https://github.com/digital-asset/daml/blob/master/ledger-api/grpc-definitions/com/daml/ledger/api/v1/command_service.proto#L24).

Reply:
Good feedback, I am adding a section on error codes to both the `participant_pruning_service.proto` (guidance to users what errors to expect) and to `PruningResult.scala` (guidance to pruning implementers on what types of errors to emit).

Reply:
> Would a new request trigger a new global consensus process, or can it use the result of a previews round?

Consensus gathering happens in the background async to  prune requests. The `eventually` with subsequent `prune` requests simply wait for the consensus to proceed beyond the specific offset in the background.